libs/auth-nest-tools/src/lib/auth.middleware.ts

```diff
@@ -18,22 +19,111 @@ interface Middleware {
   pre?(context: RequestContext): Promise<FetchParams | void>
 }
 
+export interface AuthMiddlewareOptions {
+  forwardUserInfo: boolean
+  tokenExchangeOptions?: TokenExchangeOptions
+}
+
+export interface TokenExchangeOptions {
+  issuer: string
+  clientId: string
+  clientSecret: string
+  scope: string
+}
+
 /**
  * Middleware that adds user authorization and information to OpenAPI Client requests.
  */
 export class AuthMiddleware implements Middleware {
-  constructor(private auth: Auth, private forwardUserInfo: boolean = true) {}
+  constructor(
+    private auth: Auth,
+    private options: AuthMiddlewareOptions = {
+      forwardUserInfo: true,
+    },
+    private logger?: Logger,
+  ) {}
 
   async pre(context: RequestContext) {
+    let bearerToken = this.auth.authorization
+
+    if (this.options.tokenExchangeOptions) {
+      const accessToken = await this.exchangeToken(
+        bearerToken.replace('Bearer ', ''),
+        this.options.tokenExchangeOptions,
+      )
+
+      bearerToken = `Bearer ${accessToken}`
+    }
+
     context.init.headers = Object.assign({}, context.init.headers, {
-      authorization: this.auth.authorization,
+      authorization: bearerToken,
     })
 
-    if (this.forwardUserInfo) {
+    if (this.options.forwardUserInfo) {
       context.init.headers = Object.assign({}, context.init.headers, {
         'User-Agent': this.auth.userAgent,
         'X-Real-IP': this.auth.ip,
       })
     }
   }
+
+  private async exchangeToken(
+    accessToken: string,
+    options: TokenExchangeOptions,
```

Comment:
Should access the options from `this` directly?

Reply:
Fixed.